sw/device/tests/sim_dv/pwrmgr_usbdev_smoketest.c

```diff
@@ -0,0 +1,90 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+#include "sw/device/lib/usbdev.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+static dif_pwrmgr_t pwrmgr;
+
+const test_config_t kTestConfig;
+const dif_pwrmgr_wakeup_reason_t test_wakeup_reason = {
+    .types = kDifPwrmgrWakeupTypeRequest,
+    .request_sources = kDifPwrmgrWakeupRequestSourceThree,
+};
+
+const dif_pwrmgr_wakeup_reason_t por_wakeup_reason = {
+    .types = 0,
+    .request_sources = 0,
+};
+
+static bool compare_wakeup_reasons(const dif_pwrmgr_wakeup_reason_t *lhs,
+                                   const dif_pwrmgr_wakeup_reason_t *rhs) {
+  return lhs->types == rhs->types &&
+         lhs->request_sources == rhs->request_sources;
+}
+
+// This is a DV specific pwrmgr usbdev smoketest that fakes usb activity.
+// Using the phy drive function of usbdev, we present the usbdev is in suspend
+// and thus trigger low power entry.
+// Once the system enters low power entry, the incoming USBP/USBN lines are
+// immediately different from what the wake module expects and thus causes
+// a wakeup.
+// This test just smoke checks the pwrmgr's ability to enter / exit low power
+// and the usbdev's aon wake function that has been de-coupled from the main
+// usbdev.
+
+bool test_main(void) {
+  bool test_passed = false;
+
+  CHECK(dif_pwrmgr_init(
+            (dif_pwrmgr_params_t){
+                .base_addr =
+                    mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR),
+            },
+            &pwrmgr) == kDifPwrmgrOk);
+
+  // Assuming the chip hasn't slept yet, wakeup reason should be empty.
+  dif_pwrmgr_wakeup_reason_t wakeup_reason;
+  CHECK(dif_pwrmgr_wakeup_reason_get(&pwrmgr, &wakeup_reason) == kDifPwrmgrOk);
+
+  if (compare_wakeup_reasons(&wakeup_reason, &por_wakeup_reason)) {
+    LOG_INFO("Powered up for the first time, begin test");
+  } else if (compare_wakeup_reasons(&wakeup_reason, &test_wakeup_reason)) {
+    test_passed = true;
+    LOG_INFO("USB wakeup detected");
+  } else {
+    CHECK(false);
```

Comment:
Why not just `return false;`?

Reply:
Also, adding something like `LOG_ERROR("Unexpected wakeup reason (types: %2x, sources: %8x)", wakeup_reason.types, wakeup_reason.request_sources);` before returning can be useful when debugging.

Reply:
ah i was basically using the `CHECK(false)` to give me the printout of where things failed.